<a href="https://colab.research.google.com/github/xiaann622/Analystlab/blob/main/Analyst_lab_task_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**ANALYST LAB INTERNSHIP**

• 1. E-commerce Transactions Dataset
https://www.kaggle.com/datasets/vijayuv/onlineretail

• 2. Netflix Movies & TV Shows Dataset
https://www.kaggle.com/datasets/shivamb/netflix-shows

#**  WEEK 1**
##**Task 1**

Dataset understanding

• 1. Load the dataset into Python (Pandas) or SQL.

• 2. Display the first few rows.

• 3. Identify:

• Number of rows and columns

• Data types of all columns

• Numerical features

• Categorical features

• Possible unique identifiers (primary keys)

• 4. Briefly describe what the dataset contains.


**ONLINE RETAIL STORE**

In [ ]:
import pandas as pd
import numpy as np

In [ ]:
data= pd.read_csv('/content/OnlineRetail.csv', encoding='latin1')

In [ ]:
data.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,12/1/2010 8:26,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,12/1/2010 8:26,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,12/1/2010 8:26,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,12/1/2010 8:26,3.39,17850.0,United Kingdom


In [ ]:
data.shape

(541909, 8)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    541909 non-null  object 
 1   StockCode    541909 non-null  object 
 2   Description  540455 non-null  object 
 3   Quantity     541909 non-null  int64  
 4   InvoiceDate  541909 non-null  object 
 5   UnitPrice    541909 non-null  float64
 6   CustomerID   406829 non-null  float64
 7   Country      541909 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 33.1+ MB


In [ ]:
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [ ]:
numerical_features = data.select_dtypes(include=np.number).drop(columns=['CustomerID'])
display(numerical_features.head())

,Quantity,UnitPrice
0,6,2.55
1,6,3.39
2,8,2.75
3,6,3.39
4,6,3.39


In [ ]:
categorical_features = data.select_dtypes(include='object')
display(categorical_features.head())

,InvoiceNo,StockCode,Description,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
1,536365,71053,WHITE METAL LANTERN,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,United Kingdom


In [ ]:
print(f"Unique InvoiceNo values: {data['InvoiceNo'].nunique()}")
print(f"Unique CustomerID values: {data['CustomerID'].nunique()}")
print(f"Unique InvoiceNo + StockCode combinations: {data.groupby(['InvoiceNo', 'StockCode']).size().shape[0]}")
print(f"Total number of rows: {len(data)}")

Unique InvoiceNo values: 25900
Unique CustomerID values: 4372
Unique InvoiceNo + StockCode combinations: 531225
Total number of rows: 541909


### Dataset Description

This dataset contains transactional data from an online retail store. It consists of '541,909' rows and `8' columns.

Key features include:

**Numerical Features**:

Quantity (number of items per transaction)

UnitPrice (price per item), and `CustomerID'

**Categorical Features**:

InvoiceNo

StockCode

Description
  
InvoiceDate

Country

While InvoiceNo and CustomerID are not unique for each row, a combination of InvoiceNo and StockCode appears to uniquely identify each transaction line item.

The data types indicate that InvoiceDate is currently an object and would need conversion for time-series analysis.

#**TASK 2**

**Data Cleaning**

Perform the following cleaning steps and document every action taken:

Missing Values

• Identify columns with missing values.

• Show the number of missing values per column.

• Handle missing values appropriately.

• Explain why you chose a particular method.

Duplicate Records

• Identify duplicate rows.

• Remove duplicates where necessary.

• Show the number of duplicates removed.

Standardization

Ensure consistency in:

• Date formats

• Text formatting (upper/lower case)

• Column names

• Data types

Data Validation

Identify and handle:

• Invalid values

• Inconsistent records

• Incorrect formats

• Outliers or anomalies where applicable

Cleaning Summary

Create a summary table showing:

Issue Found| Action Taken

Missing Values| Filled/Removed

Duplicates| Removed

Invalid Entries| Corrected

Standardization| Applied

In [ ]:

missing_values = data.isnull().sum()


missing_values = missing_values[missing_values > 0]

print("Columns with missing values and their counts:")
display(missing_values)

Columns with missing values and their counts:


,0
Description,1454
CustomerID,135080


In [ ]:
# Calculate the percentage of missing values
total_rows = len(data)
missing_percentage = (missing_values / total_rows) * 100

print("Percentage of missing values per column:")
display(missing_percentage.sort_values(ascending=False))

Percentage of missing values per column:


,0
CustomerID,24.926694
Description,0.268311


### Handling Missing Values

**Strategy:**

1.  **CustomerID:** As `CustomerID` is a crucial identifier for customer-centric analysis and approximately 25% of its values are missing, removing these rows is the most appropriate approach to avoid skewed analysis.
2.  **Description:** With a very small percentage (0.27%) of missing values, filling `Description` with 'Unknown' ensures all transaction records are preserved, which can be useful for broader analysis not directly dependent on product descriptions.

In [ ]:
# 1. Handling missing CustomerID values
before_drop_rows = len(data)
data.dropna(subset=['CustomerID'], inplace=True)
after_drop_rows = len(data)
dropped_customer_id_rows = before_drop_rows - after_drop_rows
print(f"Removed {dropped_customer_id_rows} rows due to missing CustomerID.")

# 2. Handling missing Description values
data['Description'].fillna('Unknown', inplace=True)
print("Filled missing 'Description' values with 'Unknown'.")


### Verification of Missing Values

Let's verify that there are no longer any missing values in the `CustomerID` and `Description` columns, and check for any other remaining missing values.

In [ ]:
missing_values_after_handling = data.isnull().sum()
missing_values_after_handling = missing_values_after_handling[missing_values_after_handling > 0]

if missing_values_after_handling.empty:
    print("No more missing values in the dataset.")
else:
    print("Remaining columns with missing values:")
    display(missing_values_after_handling)

### Duplicate Records

Let's identify and remove any duplicate rows in the dataset. A duplicate row means an identical entry across all columns. Removing them ensures that each transaction line is unique and prevents overcounting in analyses.

In [ ]:
# Identify duplicate rows
duplicate_rows = data.duplicated().sum()
print(f"Number of duplicate rows found: {duplicate_rows}")

In [ ]:
# Remove duplicate rows
before_drop_duplicates = len(data)
data.drop_duplicates(inplace=True)
after_drop_duplicates = len(data)

duplicates_removed = before_drop_duplicates - after_drop_duplicates
print(f"Number of duplicate rows removed: {duplicates_removed}")
print(f"Number of rows after removing duplicates: {len(data)}")